File: conformance/code/employee-go/internal/repository/jurisdiction_repository.go
package repository

import (
	"context"
	"time"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"gorm.io/gorm"

	"employee/internal/models"
	"employee/pkg/errors"

	tenantdb "github.com/digitnxt/digit3/src/libraries/tenant-migration/tenantdb"
	tracerobs "github.com/digitnxt/digit3/src/libraries/tracer/observability"
)

type JurisdictionRepository interface {
	Create(ctx context.Context, jurisdiction *models.Jurisdiction) error
	FindByUUID(ctx context.Context, uuid, tenantID string) (*models.Jurisdiction, error)
	// Update rewrites the mutable columns compare-and-swapping on expectedVersion.
	// The entity's Version must already be bumped to expectedVersion+1.
	Update(ctx context.Context, jurisdiction *models.Jurisdiction, expectedVersion int) error
	// DeactivateOmitted soft-deletes (is_active=false, version bumped) every
	// active jurisdiction of the employee whose id is NOT in keepIDs. Used by the
	// employee PUT/PATCH reconcile to drop jurisdictions left out of the supplied
	// array. An empty keepIDs deactivates all active jurisdictions.
	DeactivateOmitted(ctx context.Context, employeeID, tenantID, userID string, keepIDs []string) error
	Delete(ctx context.Context, id, tenantID string) error
	Search(ctx context.Context, tenantID, employeeID string, criteria *models.JurisdictionSearchCriteria) ([]*models.Jurisdiction, error)
}

type jurisdictionRepository struct {
	db *gorm.DB
}

func NewJurisdictionRepository(db *gorm.DB) JurisdictionRepository {
	return &jurisdictionRepository{db: db}
}

func (r *jurisdictionRepository) Create(ctx context.Context, jurisdiction *models.Jurisdiction) error {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.create")
	defer span.End()

	now := time.Now().UnixMilli()
	jurisdiction.CreatedTime = now
	jurisdiction.ModifiedTime = now

	start := time.Now()
	db := tenantdb.GetTenantDB(ctx, r.db)
	tx := db.WithContext(ctx).Create(jurisdiction)
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "INSERT", "jurisdictions", duration, tx.Error == nil)
	span.SetAttributes(attribute.Int64("db.duration_ms", duration.Milliseconds()))

	if tx.Error != nil {
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to create jurisdiction")
		// Translate known constraint violations to typed sentinels so the
		// service can branch via errors.Is without inspecting pg internals.
		// FK violation on employee_id surfaces as ErrJurisdictionEmployeeNotFound
		// → 404, replacing the generic 500 DATABASE_ERROR for that case.
		if translated := translatePgError(tx.Error); translated != tx.Error {
			return translated
		}
		return errors.New(errors.CodeDatabase, "failed to create jurisdiction")
	}

	span.SetStatus(codes.Ok, "Jurisdiction created")
	return nil
}

func (r *jurisdictionRepository) FindByUUID(ctx context.Context, uuid, tenantID string) (*models.Jurisdiction, error) {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.find_by_uuid")
	defer span.End()

	span.SetAttributes(attribute.String("jurisdiction.uuid", uuid))

	start := time.Now()
	var jurisdiction models.Jurisdiction
	db := tenantdb.GetTenantDB(ctx, r.db)
	tx := db.WithContext(ctx).Model(&models.Jurisdiction{}).
		Where("id = ? AND tenant_id = ?", uuid, tenantID).
		First(&jurisdiction)
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "SELECT", "jurisdictions", duration, tx.Error == nil || tx.Error == gorm.ErrRecordNotFound)
	span.SetAttributes(attribute.Int64("db.duration_ms", duration.Milliseconds()))

	if tx.Error != nil {
		if tx.Error == gorm.ErrRecordNotFound {
			span.SetStatus(codes.Ok, "Jurisdiction not found")
			return nil, errors.ErrNotFound.WithDescription("jurisdiction not found")
		}
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to find jurisdiction by UUID")
		return nil, errors.New(errors.CodeDatabase, "failed to find jurisdiction")
	}

	span.SetStatus(codes.Ok, "Jurisdiction found")
	return &jurisdiction, nil
}

func (r *jurisdictionRepository) Update(ctx context.Context, jurisdiction *models.Jurisdiction, expectedVersion int) error {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.update")
	defer span.End()

	span.SetAttributes(
		attribute.String("jurisdiction.id", jurisdiction.ID),
		attribute.String("employee.id", jurisdiction.EmployeeID),
	)

	start := time.Now()
	db := tenantdb.GetTenantDB(ctx, r.db)
	// PUT semantics — overwrite every mutable column from the fully-populated
	// entity built by UpdateJurisdictionRequest.ToEntity. Select("*") overrides
	// GORM's default zero-value-skip rule so e.g. IsActive=false is actually
	// written. Omit() guards immutable columns from accidental rewrite even
	// though ToEntity carries them forward from the loaded row.
	//
	// The WHERE pins id and version: the entity was just loaded by FindByUUID
	// and verified for employee ownership by the service, and the version
	// predicate compare-and-swaps — 0 rows means a concurrent mutation. The
	// caller must have bumped jurisdiction.Version to expectedVersion+1.
	tx := db.WithContext(ctx).Model(&models.Jurisdiction{}).
		Where("id = ? AND version = ?", jurisdiction.ID, expectedVersion).
		Select("*").
		Omit("id", "employee_id", "tenant_id", "createdBy", "createdTime").
		Updates(jurisdiction)
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "UPDATE", "jurisdictions", duration, tx.Error == nil)
	span.SetAttributes(
		attribute.Int64("db.duration_ms", duration.Milliseconds()),
		attribute.Int64("db.rows_affected", tx.RowsAffected),
	)

	if tx.Error != nil {
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to update jurisdiction")
		return errors.New(errors.CodeDatabase, "failed to update jurisdiction")
	}
	// Existence + ownership are verified by the service before this call, so a
	// zero rows-affected here means the version moved under the client → 409.
	if tx.RowsAffected == 0 {
		span.SetStatus(codes.Ok, "Jurisdiction version mismatch")
		return errors.New(errors.CodeRowVersionMismatch, "jurisdiction was modified concurrently")
	}

	span.SetStatus(codes.Ok, "Jurisdiction updated")
	return nil
}

// DeactivateOmitted soft-deletes every active jurisdiction of the employee whose
// id is not in keepIDs, in a single UPDATE (is_active=false, version bumped,
// audit set). Used by the employee reconcile to drop jurisdictions the client
// left out of the supplied array. An empty keepIDs deactivates all active rows.
func (r *jurisdictionRepository) DeactivateOmitted(ctx context.Context, employeeID, tenantID, userID string, keepIDs []string) error {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.deactivate_omitted")
	defer span.End()

	span.SetAttributes(
		attribute.String("employee.id", employeeID),
		attribute.Int("jurisdiction.keep_count", len(keepIDs)),
	)

	start := time.Now()
	db := tenantdb.GetTenantDB(ctx, r.db)
	tx := db.WithContext(ctx).Model(&models.Jurisdiction{}).
		Where("employee_id = ? AND tenant_id = ? AND is_active = ?", employeeID, tenantID, true)
	// Guard the empty-slice case: `id NOT IN ()` would render `NOT IN (NULL)`
	// (always NULL → no rows) — the opposite of "deactivate everything". Only add
	// the exclusion when there is something to keep.
	if len(keepIDs) > 0 {
		tx = tx.Where("id NOT IN ?", keepIDs)
	}
	tx = tx.Updates(map[string]interface{}{
		"is_active":    false,
		"version":      gorm.Expr("version + 1"),
		"modifiedBy":   userID,
		"modifiedTime": time.Now().UnixMilli(),
	})
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "UPDATE", "jurisdictions", duration, tx.Error == nil)
	span.SetAttributes(
		attribute.Int64("db.duration_ms", duration.Milliseconds()),
		attribute.Int64("db.rows_affected", tx.RowsAffected),
	)

	if tx.Error != nil {
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to deactivate omitted jurisdictions")
		return errors.New(errors.CodeDatabase, "failed to deactivate omitted jurisdictions")
	}

	span.SetStatus(codes.Ok, "Omitted jurisdictions deactivated")
	return nil
}

func (r *jurisdictionRepository) Delete(ctx context.Context, id, tenantID string) error {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.delete")
	defer span.End()

	span.SetAttributes(attribute.String("jurisdiction.id", id))

	start := time.Now()
	db := tenantdb.GetTenantDB(ctx, r.db)
	tx := db.WithContext(ctx).
		Where("id = ? AND tenant_id = ?", id, tenantID).
		Delete(&models.Jurisdiction{})
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "DELETE", "jurisdictions", duration, tx.Error == nil)
	span.SetAttributes(
		attribute.Int64("db.duration_ms", duration.Milliseconds()),
		attribute.Int64("db.rows_affected", tx.RowsAffected),
	)

	if tx.Error != nil {
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to delete jurisdiction")
		return errors.New(errors.CodeDatabase, "failed to delete jurisdiction")
	}
	if tx.RowsAffected == 0 {
		span.SetStatus(codes.Ok, "Jurisdiction not found")
		return errors.ErrNotFound.WithDescription("jurisdiction not found")
	}

	span.SetStatus(codes.Ok, "Jurisdiction deleted")
	return nil
}

func (r *jurisdictionRepository) Search(ctx context.Context, tenantID, employeeID string, criteria *models.JurisdictionSearchCriteria) ([]*models.Jurisdiction, error) {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.search")
	defer span.End()

	span.SetAttributes(
		attribute.String("tenant.id", tenantID),
		attribute.String("employee.id", employeeID),
	)

	var jurisdictions []*models.Jurisdiction

	start := time.Now()
	db := tenantdb.GetTenantDB(ctx, r.db)
	tx := db.WithContext(ctx).Model(&models.Jurisdiction{}).Where("tenant_id = ?", tenantID)

	// employeeID is the path-scoped owner; an empty string means "no owner
	// filter" (only used by internal callers that genuinely don't have one).
	if employeeID != "" {
		tx = tx.Where("employee_id = ?", employeeID)
	}
	if len(criteria.IDs) > 0 {
		tx = tx.Where("id IN ?", criteria.IDs)
	}
	if criteria.IsActive != nil {
		tx = tx.Where("is_active = ?", *criteria.IsActive)
	}

	tx = tx.Order("\"createdTime\" DESC")

	if criteria.Limit > 0 {
		tx = tx.Limit(criteria.Limit)
	}
	if criteria.Offset > 0 {
		tx = tx.Offset(criteria.Offset)
	}

	tx = tx.Find(&jurisdictions)
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "SELECT", "jurisdictions", duration, tx.Error == nil)
	span.SetAttributes(
		attribute.Int64("db.duration_ms", duration.Milliseconds()),
		attribute.Int("jurisdiction.found_count", len(jurisdictions)),
	)

	if tx.Error != nil {
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to search jurisdictions")
		return nil, errors.New(errors.CodeDatabase, "failed to search jurisdictions")
	}

	span.SetStatus(codes.Ok, "Jurisdictions searched")
	return jurisdictions, nil
}

func (r *jurisdictionRepository) FindByEmployeeID(ctx context.Context, employeeID, tenantID string) ([]*models.Jurisdiction, error) {
	tracer := otel.Tracer("employee-repository")
	ctx, span := tracer.Start(ctx, "db.jurisdiction.find_by_employee_id")
	defer span.End()

	span.SetAttributes(attribute.String("employee.id", employeeID))

	var jurisdictions []*models.Jurisdiction
	start := time.Now()
	db := tenantdb.GetTenantDB(ctx, r.db)
	tx := db.WithContext(ctx).Model(&models.Jurisdiction{}).
		Where("employee_id = ? AND tenant_id = ?", employeeID, tenantID).
		Find(&jurisdictions)
	duration := time.Since(start)

	tracerobs.RecordDBOperation(ctx, "SELECT", "jurisdictions", duration, tx.Error == nil)
	span.SetAttributes(
		attribute.Int64("db.duration_ms", duration.Milliseconds()),
		attribute.Int("jurisdiction.found_count", len(jurisdictions)),
	)

	if tx.Error != nil {
		span.RecordError(tx.Error)
		span.SetStatus(codes.Error, "Failed to find jurisdictions by employee ID")
		return nil, errors.New(errors.CodeDatabase, "failed to find jurisdictions by employee ID")
	}

	span.SetStatus(codes.Ok, "Jurisdictions found")
	return jurisdictions, nil
}

File: conformance/code/employee/src/main/java/com/digit/employee/repository/JurisdictionRepository.java
package com.digit.employee.repository;

import org.springframework.http.HttpStatus;

import com.digit.employee.constants.ErrorCodes;

import com.digit.employee.model.AuditDetails;
import com.digit.employee.model.BoundaryRef;
import com.digit.employee.model.Jurisdiction;
import com.digit.employee.model.JurisdictionSearchCriteria;
import org.digit.tracer.model.CustomException;
import org.digit.tracer.observability.ObservabilityMetrics;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.core.type.TypeReference;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.stereotype.Repository;

import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * JDBC repository for {@code employee_jurisdiction_v3}. Mirrors Go
 * internal/repository/jurisdiction_repository.go (GORM) using plain SQL. {@code boundary_relation} is
 * a jsonb column (serialized via Jackson). Update follows GORM {@code Updates(struct)} semantics:
 * only non-zero fields are written.
 */
@Repository
public class JurisdictionRepository {

    private static final String TABLE = "employee_jurisdiction_v3";

    private final JdbcTemplate jdbc;
    private final ObjectMapper objectMapper;
    private final ObservabilityMetrics metrics;

    public JurisdictionRepository(JdbcTemplate jdbc, ObjectMapper objectMapper, ObservabilityMetrics metrics) {
        this.jdbc = jdbc;
        this.objectMapper = objectMapper;
        this.metrics = metrics;
    }

    private static final String SELECT_COLS =
            "id, employee_id, boundary_relation, is_active, version, tenant_id, \"createdBy\", \"modifiedBy\", "
                    + "\"createdTime\", \"modifiedTime\"";

    private final RowMapper<Jurisdiction> rowMapper = (RowMapper<Jurisdiction>) (ResultSet rs, int rowNum) -> {
        Jurisdiction j = new Jurisdiction();
        Object id = rs.getObject("id");
        j.setId(id == null ? null : id.toString());
        Object empId = rs.getObject("employee_id");
        j.setEmployeeId(empId == null ? null : empId.toString());
        j.setBoundaryRelation(parseRelation(rs.getString("boundary_relation")));
        j.setActive(rs.getBoolean("is_active"));
        j.setVersion(rs.getInt("version"));
        j.setTenantId(rs.getString("tenant_id"));
        AuditDetails ad = new AuditDetails();
        ad.setCreatedBy(rs.getString("createdBy"));
        ad.setModifiedBy(rs.getString("modifiedBy"));
        ad.setCreatedTime(rs.getLong("createdTime"));
        ad.setModifiedTime(rs.getLong("modifiedTime"));
        j.setAuditDetails(ad);
        return j;
    };

    private List<BoundaryRef> parseRelation(String json) {
        if (json == null || json.isEmpty()) {
            return new ArrayList<>();
        }
        try {
            return objectMapper.readValue(json, new TypeReference<List<BoundaryRef>>() {});
        } catch (Exception e) {
            throw new RuntimeException("failed to parse boundary_relation JSON", e);
        }
    }

    private String writeRelation(List<BoundaryRef> refs) {
        try {
            return objectMapper.writeValueAsString(refs == null ? new ArrayList<>() : refs);
        } catch (Exception e) {
            throw new RuntimeException("failed to serialize boundary_relation JSON", e);
        }
    }

    /** Inserts a new jurisdiction row. The service supplies a generated UUID id. */
    public void create(Jurisdiction j) {
        long now = System.currentTimeMillis();
        j.getAuditDetails().setCreatedTime(now);
        j.getAuditDetails().setModifiedTime(now);
        if (j.getAuditDetails().getCreatedBy() == null || j.getAuditDetails().getCreatedBy().isEmpty()) {
            j.getAuditDetails().setCreatedBy("system");
        }
        if (j.getAuditDetails().getModifiedBy() == null || j.getAuditDetails().getModifiedBy().isEmpty()) {
            j.getAuditDetails().setModifiedBy("system");
        }

        boolean ok = true;
        try {
            jdbc.update(
                    "INSERT INTO " + TABLE + " (id, employee_id, boundary_relation, is_active, version, tenant_id, "
                            + "\"createdBy\", \"modifiedBy\", \"createdTime\", \"modifiedTime\") "
                            + "VALUES (?, ?, ?::jsonb, ?, ?, ?, ?, ?, ?, ?)",
                    UUID.fromString(j.getId()),
                    j.getEmployeeId() == null ? null : UUID.fromString(j.getEmployeeId()),
                    writeRelation(j.getBoundaryRelation()),
                    j.isActive(), j.getVersion(), j.getTenantId(),
                    j.getAuditDetails().getCreatedBy(), j.getAuditDetails().getModifiedBy(),
                    j.getAuditDetails().getCreatedTime(), j.getAuditDetails().getModifiedTime());
        } catch (org.springframework.dao.DataIntegrityViolationException ex) {
            // FK violation on employee_id — the client referenced an unknown employee. Mirrors Go
            // pgerr 23503 → NOT_FOUND (404) instead of a generic 500.
            ok = false;
            String m = ex.getMostSpecificCause().getMessage();
            if (m != null && (m.toLowerCase().contains("foreign key") || m.toLowerCase().contains("employee"))) {
                throw new CustomException(ErrorCodes.EMPLOYEE_NOT_FOUND, "employee not found", HttpStatus.NOT_FOUND);
            }
            throw ex;
        } catch (RuntimeException ex) {
            ok = false;
            throw ex;
        } finally {
            metrics.recordDbOperation("INSERT", "jurisdictions", ok);
        }
    }

    /** Finds a jurisdiction by id within the tenant; throws NOT_FOUND when absent. */
    public Jurisdiction findByUUID(String uuid, String tenantId) {
        boolean ok = true;
        try {
            List<Jurisdiction> rows = jdbc.query(
                    "SELECT " + SELECT_COLS + " FROM " + TABLE + " WHERE id = ? AND tenant_id = ? LIMIT 1",
                    rowMapper, UUID.fromString(uuid), tenantId);
            if (rows.isEmpty()) {
                throw new CustomException(ErrorCodes.NOT_FOUND, "The requested resource was not found", HttpStatus.NOT_FOUND);
            }
            return rows.get(0);
        } catch (CustomException ce) {
            ok = (ce.getCode() != null && ce.getCode().equals(ErrorCodes.NOT_FOUND));
            throw ce;
        } catch (RuntimeException ex) {
            ok = false;
            throw ex;
        } finally {
            metrics.recordDbOperation("SELECT", "jurisdictions", ok);
        }
    }

    /**
     * PUT overwrite of the mutable surface. Mirrors Go Update (Select("*").Omit(id, employee_id,
     * tenant_id, createdBy, createdTime)): boundary_relation and is_active are written unconditionally
     * — so {@code isActive=false} persists — while the immutable owner (employee_id) and creation
     * columns are never touched.
     *
     * <p>Optimistic concurrency: {@code version} is set to {@code expectedVersion + 1} and the WHERE
     * pins both id and the expected version. Existence + ownership are verified by the service before
     * this call, so a 0 rows-affected means the version moved → ROW_VERSION_MISMATCH (409).
     */
    public void update(Jurisdiction j, int expectedVersion) {
        long now = System.currentTimeMillis();
        String modifiedBy = j.getAuditDetails().getModifiedBy();
        if (modifiedBy == null || modifiedBy.isEmpty()) {
            modifiedBy = "system";
        }
        boolean ok = true;
        try {
            int affected = jdbc.update(
                    "UPDATE " + TABLE + " SET boundary_relation = ?::jsonb, is_active = ?, version = ?, "
                            + "\"modifiedBy\" = ?, \"modifiedTime\" = ? WHERE id = ? AND version = ?",
                    writeRelation(j.getBoundaryRelation()), j.isActive(), expectedVersion + 1, modifiedBy, now,
                    UUID.fromString(j.getId()), expectedVersion);
            if (affected == 0) {
                throw new CustomException(ErrorCodes.ROW_VERSION_MISMATCH, "jurisdiction was modified concurrently", HttpStatus.CONFLICT);
            }
        } catch (CustomException ce) {
            ok = ErrorCodes.ROW_VERSION_MISMATCH.equals(ce.getCode());
            throw ce;
        } catch (RuntimeException ex) {
            ok = false;
            throw ex;
        } finally {
            metrics.recordDbOperation("UPDATE", "jurisdictions", ok);
        }
    }

    /**
     * Soft-deletes (is_active=false, version bumped, audit set) every active jurisdiction of the
     * employee whose id is NOT in {@code keepIds}, in a single UPDATE. Used by the employee PUT/PATCH
     * reconcile to drop jurisdictions the client left out of the supplied array. An empty
     * {@code keepIds} deactivates every active jurisdiction. Mirrors Go DeactivateOmitted.
     */
    public void deactivateOmitted(String employeeId, String tenantId, String userId, List<String> keepIds) {
        long now = System.currentTimeMillis();
        String modifiedBy = (userId == null || userId.isEmpty()) ? "system" : userId;
        StringBuilder sql = new StringBuilder(
                "UPDATE " + TABLE + " SET is_active = false, version = version + 1, "
                        + "\"modifiedBy\" = ?, \"modifiedTime\" = ? "
                        + "WHERE employee_id = ? AND tenant_id = ? AND is_active = true");
        List<Object> args = new ArrayList<>();
        args.add(modifiedBy);
        args.add(now);
        args.add(UUID.fromString(employeeId));
        args.add(tenantId);
        // Guard the empty case: an empty NOT IN (...) is invalid SQL and would also mean "keep nothing",
        // so only add the exclusion when there is something to keep.
        if (keepIds != null && !keepIds.isEmpty()) {
            sql.append(" AND id NOT IN (").append(placeholders(keepIds.size())).append(")");
            for (String k : keepIds) { args.add(UUID.fromString(k)); }
        }
        boolean ok = true;
        try {
            jdbc.update(sql.toString(), args.toArray());
        } catch (RuntimeException ex) {
            ok = false;
            throw ex;
        } finally {
            metrics.recordDbOperation("UPDATE", "jurisdictions", ok);
        }
    }

    /** Hard-deletes a jurisdiction by id within the tenant; throws NOT_FOUND when no row matches. */
    public void delete(String id, String tenantId) {
        boolean ok = true;
        try {
            int affected = jdbc.update("DELETE FROM " + TABLE + " WHERE id = ? AND tenant_id = ?",
                    UUID.fromString(id), tenantId);
            if (affected == 0) {
                throw new CustomException(ErrorCodes.NOT_FOUND, "The requested resource was not found", HttpStatus.NOT_FOUND);
            }
        } catch (CustomException ce) {
            ok = (ce.getCode() != null && ce.getCode().equals(ErrorCodes.NOT_FOUND));
            throw ce;
        } catch (RuntimeException ex) {
            ok = false;
            throw ex;
        } finally {
            metrics.recordDbOperation("DELETE", "jurisdictions", ok);
        }
    }

    /**
     * Searches jurisdictions scoped to an owning employee. Mirrors Go
     * Search(ctx, tenantID, employeeID, criteria) — employeeId is a positional owner scope from the
     * nested URL path, not a criteria field.
     */
    public List<Jurisdiction> search(String tenantId, String employeeId, JurisdictionSearchCriteria c) {
        StringBuilder sql = new StringBuilder("SELECT " + SELECT_COLS + " FROM " + TABLE + " WHERE 1=1");
        List<Object> args = new ArrayList<>();

        if (tenantId != null && !tenantId.isEmpty()) {
            sql.append(" AND tenant_id = ?");
            args.add(tenantId);
        }
        if (employeeId != null && !employeeId.isEmpty()) {
            sql.append(" AND employee_id = ?");
            args.add(UUID.fromString(employeeId));
        }
        if (c.getIds() != null && !c.getIds().isEmpty()) {
            sql.append(" AND id IN (").append(placeholders(c.getIds().size())).append(")");
            for (String s : c.getIds()) { args.add(UUID.fromString(s)); }
        }
        if (c.getIsActive() != null) {
            sql.append(" AND is_active = ?");
            args.add(c.getIsActive());
        }

        // Fixed server-side ordering, matching Go (post-8749c30e removed client-controlled sort).
        // Never concatenate client sort input into SQL — that was an injection sink.
        sql.append(" ORDER BY \"createdTime\" DESC");

        if (c.getLimit() > 0) {
            sql.append(" LIMIT ?");
            args.add(c.getLimit());
        }
        if (c.getOffset() > 0) {
            sql.append(" OFFSET ?");
            args.add(c.getOffset());
        }

        boolean ok = true;
        try {
            return jdbc.query(sql.toString(), rowMapper, args.toArray());
        } catch (RuntimeException ex) {
            ok = false;
            throw ex;
        } finally {
            metrics.recordDbOperation("SELECT", "jurisdictions", ok);
        }
    }

    private static boolean notEmpty(String s) {
        return s != null && !s.isEmpty();
    }

    private static String placeholders(int n) {
        List<String> p = new ArrayList<>();
        for (int i = 0; i < n; i++) { p.add("?"); }
        return p.stream().collect(Collectors.joining(", "));
    }
}
